drivers/ad5770r/ad5770r.c

```diff
@@ -0,0 +1,658 @@
+	memcpy(&data[1], reg_data, count);
+
+	return spi_write_and_read(dev->spi_desc, data, count + 1);
+}
+
+/**
+ * SPI write to device using a mask.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - The mask.
+ * @param data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_spi_write_mask(struct ad5770r_dev *dev,
+			       uint8_t reg_addr,
+			       uint32_t mask,
+			       uint8_t data)
+{
+	uint8_t reg_data;
+	int32_t ret;
+
+	ret = ad5770r_spi_reg_read(dev, reg_addr, &reg_data);
+	reg_data &= ~mask;
+	reg_data |= data;
+	ret |= ad5770r_spi_reg_write(dev, reg_addr, reg_data);
+
+	return ret;
+}
+
+/**
+ * Set device spi settings.
+ * @param dev - The device structure.
+ * @param spi_settings - The structure that contains the device spi
+ *		       parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_device_spi(struct ad5770r_dev *dev,
+			       const struct ad5770r_device_spi_settings *spi_settings)
+{
+	int32_t ret;
+
+	ret = ad5770r_spi_write_mask(dev,
+				     AD5770R_INTERFACE_CONFIG_A,
+				     AD5770R_INTERFACE_CONFIG_A_ADDR_ASCENSION_MSB_MSK,
+				     AD5770R_INTERFACE_CONFIG_A_ADDR_ASCENSION_MSB(spi_settings->addr_ascension));
+
+	ret |= ad5770r_spi_write_mask(dev,
+				      AD5770R_INTERFACE_CONFIG_B,
+				      AD5770R_INTERFACE_CONFIG_B_SINGLE_INST_MSK,
+				      AD5770R_INTERFACE_CONFIG_B_SINGLE_INST(spi_settings->single_instruction));
+
+	ret |= ad5770r_spi_reg_write(dev,
+				     AD5770R_STREAM_MODE,
+				     spi_settings->stream_mode_length);
+
+	if (ret)
+		return ret;
+
+	dev->dev_spi_settings.addr_ascension =
+		spi_settings->addr_ascension;
+	dev->dev_spi_settings.single_instruction =
+		spi_settings->single_instruction;
+	dev->dev_spi_settings.stream_mode_length =
+		spi_settings->stream_mode_length;
+
+	return SUCCESS;
+}
+
+/**
+ * Set the channel configuration.
+ * @param dev - The device structure.
+ * @param channel_config - The structure that contains the channel
+ *			configuration.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_channel_config(struct ad5770r_dev *dev,
+			       const struct ad5770r_channel_switches *channel_config)
+{
+	int32_t ret;
+
+	ret = ad5770r_spi_reg_write(dev,
+				    AD5770R_CHANNEL_CONFIG,
+				    AD5770R_CHANNEL_CONFIG_CH0_SHUTDOWN_B(channel_config->en0) |
+				    AD5770R_CHANNEL_CONFIG_CH1_SHUTDOWN_B(channel_config->en1) |
+				    AD5770R_CHANNEL_CONFIG_CH2_SHUTDOWN_B(channel_config->en2) |
+				    AD5770R_CHANNEL_CONFIG_CH3_SHUTDOWN_B(channel_config->en3) |
+				    AD5770R_CHANNEL_CONFIG_CH4_SHUTDOWN_B(channel_config->en4) |
+				    AD5770R_CHANNEL_CONFIG_CH5_SHUTDOWN_B(channel_config->en5) |
+				    AD5770R_CHANNEL_CONFIG_CH0_SINK_EN(channel_config->sink0));
+
+	if (ret)
+		return ret;
+
+	dev->channel_config = *channel_config;
```

Comment:
this still passes a reference to an external object, which may have limited life-time

Reply:
wait; my bad
it de-references the object;

Reply:
it would be a good idea to do `memcpy(&dev->channel_config, channel_config, sizeof(dev->channel_config))`
this is more certain to work well on most C compilers;

Reply:
No, keep it the way it is. This should work with all compilers either way.